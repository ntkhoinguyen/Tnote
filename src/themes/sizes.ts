const padding = {
  xs: 4,
  sm: 8,
  md: 12,
  lg: 16,
  xl: 20,
  xxl: 24,
};

const margin = {
  xs: 4,
  sm: 8,
  md: 12,
  lg: 16,
  xl: 20,
  xxl: 24,
};

const borderRadius = {
  xs: 4,
  sm: 8,
  md: 12,
  lg: 16,
  xl: 20,
  xxl: 24,
};

const fontSize = {
  xs: 10,
  sm: 12,
  md: 14,
  lg: 16,
  xl: 18,
  xxl: 20,
};

const spacing = {
  xs: 4,
  sm: 8,
  md: 12,
  lg: 16,
  xl: 20,
  xxl: 24,
};

const fontWeight = {
  regular: "400",
  medium: "500",
  bold: "700",
};

const lineHeight = {
  xs: 10,
  sm: 12,
  md: 14,
  lg: 16,
  xl: 18,
  xxl: 20,
};

const borderWidth = {
  xs: 1,
  sm: 2,
  md: 3,
  lg: 4,
  xl: 5,
  xxl: 6,
};

const heightHeader = 80;
const heightHeaderModal = 40;
const heightCreateButton = 40;;

const appLogoWidth = 100;
const appLogoHeight = 100;

export const NUM_BARS = 13;

export const indicatorSize = {
  verySmall: {
    width: 3,
    height: 7,
    borderRadius: 2,
    translateY: -10,
  },
  small: {
    width: 3,
    height: 7,
    borderRadius: 2,
    translateY: -15,
  },
  medium: {
    width: 5,
    height: 17,
    borderRadius: 3,
    translateY: -33,
  },
  large: {
    width: 8,
    height: 20,
    borderRadius: 3,
    translateY: -40,
  },
};

export const sizes = {
  padding,
  margin,
  borderRadius,
  fontSize,
  spacing,
  fontWeight,
  lineHeight,
  borderWidth,
  heightHeader,
  heightHeaderModal,
  heightCreateButton,
  appLogoWidth,
  appLogoHeight,
};
