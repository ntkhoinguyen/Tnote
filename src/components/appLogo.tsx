import React, { useEffect, memo } from "react";
import { StyleSheet } from "react-native";
import Animated, {
  useSharedValue,
  useAnimatedStyle,
  Easing,
  withRepeat,
  withTiming,
} from "react-native-reanimated";

import { sizes } from "@/src/themes/sizes";
import { AppLogoType } from "@/src/utils/types";

export const AppLogo: React.FC<AppLogoType> = memo((props) => {
  const { isEffect } = props;
  const opacity = useSharedValue(isEffect ? 0.3 : 1);

  useEffect(() => {
    // Thực hiện hiệu ứng nhấp nháy với withRepeat
    if (isEffect) {
      opacity.value = withRepeat(
        withTiming(1, { duration: 1000, easing: Easing.ease }), // Tăng sáng
        -1, // Lặp lại vô hạn
        true // Lặp lại từ 1 sang 0 (từ sáng đến tối)
      );
    }
  }, [opacity, isEffect]);

  const aniStyle = useAnimatedStyle(() => {
    return {
      opacity: opacity.value,
    };
  });

  return (
    <Animated.Image
      testID="imgAppLogo"
      resizeMode="contain"
      source={require("@/assets/images/logo.png")}
      style={[styles.logo, aniStyle]}
    />
  );
});

AppLogo.displayName = "AppLogo";

const styles = StyleSheet.create({
  logo: {
    width: sizes.appLogoWidth,
    height: sizes.appLogoHeight,
  },
});
