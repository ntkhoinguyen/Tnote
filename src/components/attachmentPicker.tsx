import React, { useMemo, useState } from "react";
import {
  View,
  Text,
  StyleSheet,
  TouchableOpacity,
  Alert,
  Image,
} from "react-native";
import { CameraView, CameraType, useCameraPermissions } from "expo-camera";
import * as DocumentPicker from "expo-document-picker";

import { useAppContext } from "@/src/useHook/useAppContext";
import { defaultColors } from "@/src/themes/colors";
import { sizes } from "@/src/themes/sizes";
import {
  FontAwesomeIcon,
  MaterialIconsIcon,
  FontAwesome6Icon,
} from "@/src/components/icon";
import {
  AttachmentPickerType,
  OptionImageType,
  RenderCameraType,
} from "@/src/utils/types";
import { useModal } from "@/src/useHook/useModal";

export const AttachmentPicker = (props: AttachmentPickerType) => {
  const { colors, sizes, t } = useAppContext();
  const styles = useMemo(() => createStyles(colors, sizes), [colors, sizes]);

  const { multiple = true, type = "*/*", RenderItem } = props;

  const { open, close, RenderModal } = useModal({
    title: t("selectImage"),
    animationType: "slide",
    containerStyle: { justifyContent: "flex-end" },
    contentStyle: { height: 300 },
    content: (props: any) => (
      <RenderOptionImage onImage={onImage} onFolder={onFolder} />
    ),
  });

  const {
    open: openImage,
    close: closeImage,
    RenderModal: RenderModalImage,
  } = useModal({
    title: t("takeAPhoto"),
    animationType: "slide",
    contentStyle: { height: "100%" },
    content: (props: any) => <RenderCamera onSelect={onSelectImage} />,
  });

  const onPress = () => {
    open();
  };

  const onImage = () => {
    close();
    openImage();
  };

  const pickDocument = async () => {
    try {
      const result = await DocumentPicker.getDocumentAsync({
        type: type,
        multiple: multiple,
        copyToCacheDirectory: true,
      });
      if (result.assets && result.assets.length > 0 && !result.canceled) {
        const files = result.assets.map((asset) => asset.uri);
        props.onSelect(files, false);
      }
    } catch (error) {
      console.log("[attachmentPicker][pickDocument] ----> ", error);
    }
  };

  const onFolder = () => {
    close();
    pickDocument();
  };

  const onSelectImage = (uri: string) => {
    closeImage();
    props.onSelect([uri], true);
  };

  return (
    <TouchableOpacity
      testID="ModalInputSelection"
      onPress={onPress}
      style={styles.container}
    >
      {RenderItem ? (
        <RenderItem />
      ) : (
        <FontAwesomeIcon name="camera" size={14} color={colors.gray} />
      )}

      {RenderModal()}
      {RenderModalImage()}
    </TouchableOpacity>
  );
};

const RenderOptionImage = (props: OptionImageType) => {
  const { colors, sizes, t } = useAppContext();
  const styles = useMemo(() => createStyles(colors, sizes), [colors, sizes]);

  const onImagePress = () => {
    props.onImage();
  };

  const onFolderPress = () => {
    props.onFolder();
  };

  return (
    <View style={styles.OptionsContainer}>
      <TouchableOpacity style={styles.optionItemContent} onPress={onImagePress}>
        <View style={styles.optionItem}>
          <FontAwesomeIcon name="camera" size={40} color={colors.gray} />
        </View>
        <Text style={styles.optionItemText}>{t("camera")}</Text>
      </TouchableOpacity>

      <TouchableOpacity
        style={styles.optionItemContent}
        onPress={onFolderPress}
      >
        <View style={styles.optionItem}>
          <FontAwesomeIcon name="folder" size={40} color={colors.gray} />
        </View>
        <Text style={styles.optionItemText}>{t("folder")}</Text>
      </TouchableOpacity>
    </View>
  );
};

const RenderCamera = (props: RenderCameraType) => {
  const { colors, sizes, t } = useAppContext();
  const styles = useMemo(() => createStyles(colors, sizes), [colors, sizes]);

  const [permission, requestPermission] = useCameraPermissions();

  const [facing, setFacing] = useState<CameraType>("back");
  const [flashMode, setFlashMode] = useState<"on" | "off" | "auto">("off");
  const [uri, setUri] = useState<string | null>(null);

  const cameraRef = React.useRef<CameraView>(null);

  const toggleCameraFacing = () => {
    if (facing === "back") {
      setFlashMode("off");
    }
    setFacing((current) => (current === "back" ? "front" : "back"));
  };

  const toggleFlashMode = () => {
    setFlashMode((current) => (current === "on" ? "off" : "on"));
  };

  const onCancel = () => {
    setUri(null);
  };

  const onSave = () => {
    props.onSelect(uri || "");
  };

  if (!permission) {
    // Camera permissions are still loading.
    return <View />;
  }

  if (!permission.granted) {
    // Camera permissions are not granted yet.
    return (
      <View style={styles.permissionContainer}>
        <Text style={styles.permissionText}>{t("needPermissionCamera")}</Text>
        <TouchableOpacity onPress={requestPermission}>
          <Text style={styles.grantPermission}>{t("grantPermission")}</Text>
        </TouchableOpacity>
      </View>
    );
  }

  const onTakeAPhoto = async () => {
    try {
      if (cameraRef.current) {
        const result = await cameraRef.current?.takePictureAsync?.({
          shutterSound: false,
        });
        if (result.uri) {
          setUri(result.uri);
        } else {
          Alert.alert(t("cannotTakeAPicture"));
        }
      } else {
        Alert.alert(t("cannotTakeAPicture"));
      }
    } catch (error) {
      Alert.alert(t("cannotTakeAPicture"));
      console.error("[camera][onTakeAPhoto] ----", error);
    }
  };

  return (
    <View style={styles.CameraContainer}>
      {uri ? (
        <View style={styles.imageContainer}>
          <Image
            source={{ uri: uri }}
            style={styles.image}
            resizeMode="cover"
          />
          <View style={styles.actionCancelContent}>
            <TouchableOpacity onPress={onCancel}>
              <MaterialIconsIcon name="cancel" size={32} color={colors.gray} />
            </TouchableOpacity>
          </View>
          <View style={styles.actionSaveContent}>
            <TouchableOpacity onPress={onSave}>
              <FontAwesome6Icon
                name="circle-check"
                size={40}
                color={colors.white}
              />
            </TouchableOpacity>
          </View>
        </View>
      ) : (
        <CameraView
          ref={cameraRef}
          style={styles.camera}
          facing={facing}
          flash={flashMode}
          animateShutter={false}
        >
          <View style={styles.buttonContainer}>
            <View style={styles.rightButtons}>
              <TouchableOpacity
                style={styles.button}
                onPress={toggleCameraFacing}
              >
                <MaterialIconsIcon
                  name="flip-camera-ios"
                  size={32}
                  color={colors.white}
                />
              </TouchableOpacity>
              <TouchableOpacity
                style={styles.button}
                onPress={toggleFlashMode}
                disabled={facing === "front"}
              >
                <MaterialIconsIcon
                  name={flashMode === "on" ? "flash-on" : "flash-off"}
                  size={32}
                  color={colors.white}
                />
              </TouchableOpacity>
            </View>

            <TouchableOpacity style={styles.buttonTake} onPress={onTakeAPhoto}>
              <FontAwesomeIcon name="camera" size={32} color={colors.white} />
            </TouchableOpacity>
          </View>
        </CameraView>
      )}
    </View>
  );
};

const createStyles = (colors: typeof defaultColors, size: typeof sizes) => {
  return StyleSheet.create({
    container: {},
    text: {
      fontWeight: size.fontWeight.medium as "500",
    },
    OptionsContainer: {
      flex: 1,
      flexDirection: "row",
      alignItems: "center",
      justifyContent: "space-around",
      paddingVertical: size.padding.sm,
      gap: size.margin.xxl,
    },
    optionItemContent: {
      alignItems: "center",
      justifyContent: "center",
      gap: size.margin.sm,
    },
    optionItem: {
      borderWidth: size.borderWidth.xs,
      borderRadius: size.borderRadius.md,
      borderColor: colors.border,
      padding: size.padding.md,
      alignItems: "center",
      justifyContent: "center",
      backgroundColor: colors.light,
    },
    optionItemText: {
      fontSize: size.fontSize.lg,
      color: colors.text,
    },
    permissionContainer: {
      flex: 1,
      alignItems: "center",
      justifyContent: "center",
      gap: size.margin.lg,
    },
    permissionText: {
      fontSize: size.fontSize.xl,
      color: colors.orange,
    },
    grantPermission: {
      padding: size.padding.md,
      borderWidth: size.borderWidth.xs,
      borderRadius: size.borderRadius.xs,
      borderColor: colors.border,
      backgroundColor: colors.primary,
      color: colors.white,
      fontWeight: size.fontWeight.bold as "bold",
      fontSize: size.fontSize.md,
    },
    CameraContainer: {
      flex: 1,
    },

    camera: {
      flex: 1,
    },
    buttonContainer: {
      flex: 1,
      backgroundColor: "transparent",
      margin: sizes.margin.lg,
      alignItems: "center",
    },
    rightButtons: {
      flex: 1,
      width: "100%",
      alignItems: "flex-end",
    },
    button: {
      marginTop: sizes.margin.xl,
    },
    buttonTake: {
      alignSelf: "center",
      marginBottom: sizes.margin.xxl,
    },
    imageContainer: {
      flex: 1,
    },
    actionImage: {
      flexDirection: "row",
      position: "absolute",
      bottom: 0,
      width: "100%",
      justifyContent: "center",
      gap: sizes.margin.xxl * 3,
    },
    image: {
      height: "100%",
      width: "100%",
    },
    actionSaveContent: {
      position: "absolute",
      width: "100%",
      alignItems: "center",
      bottom: sizes.margin.lg,
    },
    actionCancelContent: {
      position: "absolute",
      width: "100%",
      top: 0,
      alignItems: "flex-end",
      padding: sizes.padding.sm,
    },
  });
};
