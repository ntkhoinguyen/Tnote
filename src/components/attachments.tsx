import React, { useMemo } from "react";
import { View, StyleSheet, Image, Text, TouchableOpacity } from "react-native";

import { useAppContext } from "@/src/useHook/useAppContext";
import { defaultColors } from "@/src/themes/colors";
import { sizes } from "@/src/themes/sizes";
import { AttachmentsType, AttachmentType } from "@/src/utils/types";
import { imageExtensions } from "../utils/utils";
import {
  IoniconsIcon,
  FoundationIcon,
  AntDesignIcon,
  FontAwesomeIcon,
} from "./icon";

export const fileIcons = {
  pdf: (
    <AntDesignIcon
      name="pdffile1"
      size={24}
      color={defaultColors.red}
      style={{ marginTop: sizes.margin.xs }}
    />
  ),
  doc: (
    <FoundationIcon
      name="page-doc"
      size={24}
      color={defaultColors.primary}
      style={{ marginTop: sizes.margin.xs }}
    />
  ),
  docx: (
    <FoundationIcon
      name="page-doc"
      size={24}
      color={defaultColors.primary}
      style={{ marginTop: sizes.margin.xs }}
    />
  ),
  document: (
    <IoniconsIcon
      name="document"
      size={24}
      color={defaultColors.yellow}
      style={{ marginTop: sizes.margin.xs }}
    />
  ),
};

export const Attachments = (props: AttachmentsType) => {
  const { colors, sizes } = useAppContext();
  const styles = useMemo(() => createStyles(colors, sizes), [colors, sizes]);

  const { uri, onPreview, onDelete } = props;

  return (
    <View style={styles.container}>
      {(uri || []).map((item, index) => {
        return (
          <RenderAttachmentItem
            uri={item}
            key={"attachment" + index}
            onPreview={onPreview}
            onDelete={onDelete}
          />
        );
      })}
    </View>
  );
};

const RenderAttachmentItem = (props: AttachmentType) => {
  const { uri, onPreview, onDelete } = props;

  const { colors, sizes } = useAppContext();
  const styles = useMemo(() => createStyles(colors, sizes), [colors, sizes]);

  const fileName = useMemo(() => {
    const split = uri.split("/");
    return split[split.length - 1];
  }, [uri]);

  const fileType = useMemo(() => {
    const split = uri.split(".");
    return split[split.length - 1].toLowerCase();
  }, [uri]);

  const isImage = useMemo(() => {
    return imageExtensions.includes(fileType);
  }, [fileType]);

  const fileIcon = useMemo(() => {
    return (
      fileIcons[fileType as keyof typeof fileIcons] || fileIcons["document"]
    );
  }, [fileType]);

  const onPreviewAttachment = () => {
    onPreview?.(uri);
  };

  const onDeleteAttachment = () => {
    onDelete?.(uri);
  };

  return (
    <TouchableOpacity
      style={styles.attachmentContainer}
      onPress={onPreviewAttachment}
    >
      {isImage ? (
        <Image source={{ uri: uri }} style={styles.image} />
      ) : (
        <View style={styles.IconContent}>
          {fileIcon}
          <View style={{ flex: 1 }}>
            <Text numberOfLines={3} style={styles.text}>
              {fileName}
            </Text>
          </View>
        </View>
      )}
      {onDelete ? (
        <TouchableOpacity style={styles.delete} onPress={onDeleteAttachment}>
          <FontAwesomeIcon name="close" size={24} color={colors.red} />
        </TouchableOpacity>
      ) : null}
    </TouchableOpacity>
  );
};

const createStyles = (colors: typeof defaultColors, size: typeof sizes) => {
  return StyleSheet.create({
    container: {
      flexWrap: "wrap",
      flexDirection: "row",
      width: "100%",
      gap: sizes.margin.sm,
    },
    attachmentContainer: {
      borderRadius: size.borderRadius.sm,
      borderColor: colors.border,
      borderWidth: sizes.borderWidth.xs,
      padding: sizes.padding.sm,
    },
    image: {
      height: 50,
      width: 50,
    },
    IconContent: {
      height: 50,
      width: 150,
      flexDirection: "row",
      backgroundColor: colors.background,
    },
    text: {
      fontSize: size.fontSize.md,
      fontWeight: size.fontWeight.bold as "bold",
      color: colors.text,
      marginLeft: size.margin.sm,
      marginRight: size.margin.lg,
    },
    delete: {
      position: "absolute",
      top: 0,
      right: sizes.margin.xs,
    },
  });
};
