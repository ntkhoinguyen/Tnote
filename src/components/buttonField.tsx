// components/buttonField.tsx
import React, { useMemo, useRef, memo } from "react";
import { TouchableOpacity, Text, StyleSheet } from "react-native";
import { useAppContext } from "@/src/useHook/useAppContext";
import { defaultColors } from "../themes/colors";
import { sizes } from "@/src/themes/sizes";
import { ButtonFieldType } from "../utils/types";
import { getButtonStyles } from "@/src/utils/utils";

export const ButtonField: React.FC<ButtonFieldType> = memo((props) => {
  const { colors, sizes } = useAppContext();
  const {
    text,
    disabled,
    type = "fill",
    color = colors.primary,
    containerStyle = {},
    LeftSection,
    RightSection,
    onPress,
  } = props;

  const styles = useMemo(() => createStyles(colors, sizes), [colors, sizes]);
  const timer = useRef<number | null>(null);
  const { backgroundColor, textColor, borderColor } = useMemo(
    () => getButtonStyles(type, color, colors),
    [type, color, colors]
  );

  const onPressHandler = () => {
    if (timer.current) return;
    if (!disabled && onPress) onPress();

    timer.current = setTimeout(() => {
      timer.current = null;
    }, 2000);
  };

  return (
    <TouchableOpacity
      style={[styles.button, { backgroundColor, borderColor }, containerStyle]}
      disabled={disabled}
      onPress={onPressHandler}
      accessibilityRole="button"
      accessibilityLabel="ButtonField"
    >
      {LeftSection ? <LeftSection /> : null}

      {text ? (
        <Text style={[styles.text, { color: textColor }]}>{text}</Text>
      ) : null}

      {RightSection ? <RightSection /> : null}
    </TouchableOpacity>
  );
});

ButtonField.displayName = "ButtonField";

const createStyles = (colors: typeof defaultColors, size: typeof sizes) =>
  StyleSheet.create({
    button: {
      padding: size.padding.xs,
      borderRadius: size.borderRadius.sm,
      marginVertical: size.margin.sm,
      flexDirection: "row",
      borderWidth: size.borderWidth.xs,
      justifyContent: "space-around",
    },
    text: {
      color: colors.white,
      textAlign: "center",
      fontWeight: size.fontWeight.bold as "bold",
      fontSize: size.fontSize.lg,
      marginHorizontal: size.margin.sm,
    },
  });
