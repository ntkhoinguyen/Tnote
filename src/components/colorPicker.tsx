import React, { useCallback, useMemo, useRef } from "react";
import { TouchableOpacity, View, StyleSheet } from "react-native";
import { FlashList } from "@shopify/flash-list";

import { useAppContext } from "@/src/useHook/useAppContext";
import { defaultColors, colorsPicker } from "@/src/themes/colors";
import { sizes } from "@/src/themes/sizes";
import { usePopover } from "@/src/useHook/usePopover";
import { PopoverWrapper } from "@/src/components/popoverWrapper";
import { ColorPickerType } from "../utils/types";

export const ColorPicker: React.FC<ColorPickerType> = (props) => {
  const { colors, sizes } = useAppContext();

  const styles = useMemo(
    () => createStyles(colors, sizes, { backgroundColor: props.value }),
    [colors, sizes, props.value]
  );
  const colorRef = useRef(null);
  const { visible, target, content, open, close } = usePopover();

  const onSelectColor = (color: string) => {
    props.onSelect(color);
    close();
  };
  const onOpenPicker = () => {
    open({
      targetRef: colorRef,
      content: <ListColorPicker onSelect={onSelectColor} />,
    });
  };

  const onClosedPopover = () => {
    close();
  };

  return (
    <View>
      <TouchableOpacity testID="colorPickerBtn" onPress={onOpenPicker}>
        <View ref={colorRef} style={styles.container} />
      </TouchableOpacity>
      <PopoverWrapper
        visible={visible}
        target={target}
        onClose={onClosedPopover}
        arrowSize={{ width: 12, height: 12 }}
        popoverStyle={styles.popoverStyle}
      >
        {content}
      </PopoverWrapper>
    </View>
  );
};

type OptionsComponetType = {
  onSelect: (color: string) => void;
};

export const ListColorPicker = (props: OptionsComponetType) => {
  const { colors, sizes } = useAppContext();
  const styles = useMemo(
    () => createStyles(colors, sizes, {}),
    [colors, sizes]
  );

  const renderItem = useCallback(
    ({ item, index }: { item: string; index: number }) => {
      const onSelectColor = () => props.onSelect(item);
      return (
        <TouchableOpacity
          key={index}
          testID={`itemColor${index}`}
          onPress={onSelectColor}
        >
          <View style={[styles.item, { backgroundColor: item }]} />
        </TouchableOpacity>
      );
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );

  return (
    <View testID="colorPickerList" style={styles.colorContainer}>
      <FlashList
        data={colorsPicker}
        renderItem={renderItem}
        keyExtractor={(item) => item}
        numColumns={4}
      />
    </View>
  );
};

type paramsStyleType = {
  backgroundColor?: string;
};

const createStyles = (
  colors: typeof defaultColors,
  size: typeof sizes,
  params: paramsStyleType
) => {
  const { backgroundColor } = params;
  return StyleSheet.create({
    container: {
      backgroundColor: backgroundColor || colors.backgroundPopup,
      borderRadius: size.borderRadius.xs,
      height: 24,
      width: 24,
      borderWidth: size.borderWidth.xs,
      borderColor: colors.border,
    },
    popoverStyle: {
      backgroundColor: colors.backgroundPopup,
    },
    colorContainer: {
      width: 150,
      padding: size.padding.sm,
      backgroundColor: colors.backgroundPopup,
    },
    item: {
      borderRadius: size.borderRadius.xs,
      height: 24,
      width: 24,
      margin: size.margin.xs,
      borderWidth: size.borderWidth.xs,
      borderColor: colors.border,
    },
  });
};
