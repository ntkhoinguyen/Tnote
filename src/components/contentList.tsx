import React, { useCallback, useMemo, useState } from "react";
import { View, StyleSheet, TouchableOpacity, Text } from "react-native";
import { FlashList } from "@shopify/flash-list";

import { useAppContext } from "@/src/useHook/useAppContext";
import { defaultColors } from "../themes/colors";
import { sizes } from "@/src/themes/sizes";
import { normalize } from "../utils/utils";

import { SearchBar } from "@/src/components/searchBar";
import { FontAwesome5Icon, FontAwesome6Icon } from "@/src/components/icon";

import {
  ContentListType,
  ModalInputSelectionItemType,
  RenderItemDefaultProps,
} from "@/src/utils/types";

export const ContentList = (props: ContentListType) => {
  const { haveSearch, options, haveAddNew, haveUnSelect, RenderItemCustom } =
    props;
  const { colors, sizes } = useAppContext();
  const styles = useMemo(() => createStyles(colors, sizes), [colors, sizes]);

  const [searchText, setSearchText] = useState("");

  const optionFilter = useMemo(
    () =>
      options.filter((item) =>
        normalize(item.title.toLowerCase().trim()).includes(
          normalize(searchText.toLowerCase().trim())
        )
      ),
    [options, searchText]
  );

  const onAddNew = () => {
    props.onAddNew?.();
    props.onClose?.();
  };

  const renderItem = ({ item }: { item: ModalInputSelectionItemType }) => {
    if (RenderItemCustom)
      return (
        <RenderItemCustom
          key={item.id}
          item={item as any}
          selected={props.selected}
          haveUnSelect={haveUnSelect}
          onSelect={props.onSelect}
          onUnSelect={props.onUnSelect}
          onClose={props.onClose}
        />
      );
    return (
      <RenderItemDefault
        key={item.id}
        item={item}
        selected={props.selected}
        haveUnSelect={haveUnSelect}
        onSelect={props.onSelect}
        onUnSelect={props.onUnSelect}
        onClose={props.onClose}
      />
    );
  };

  const itemSeparatorComponent = useCallback(() => {
    return <View style={styles.separator} />;
  }, [styles.separator]);

  return (
    <View testID="contentList" style={styles.container}>
      {haveSearch ? (
        <View style={{ height: 40, width: "100%" }}>
          <SearchBar onSearch={setSearchText} />
        </View>
      ) : null}
      <View style={styles.content}>
        <FlashList
          data={optionFilter}
          keyExtractor={(item) => item.id.toString()}
          renderItem={renderItem}
          ItemSeparatorComponent={itemSeparatorComponent}
          showsVerticalScrollIndicator={true}
        />
      </View>
      {haveAddNew ? (
        <TouchableOpacity style={styles.btnAddItem} onPress={onAddNew}>
          <Text style={styles.textAddItem}>+ Thêm mới</Text>
        </TouchableOpacity>
      ) : null}
    </View>
  );
};

export const RenderItemDefault = (props: RenderItemDefaultProps) => {
  const { colors, sizes } = useAppContext();
  const styles = useMemo(() => createStyles(colors, sizes), [colors, sizes]);

  const { item, selected, haveUnSelect, onSelect, onUnSelect, onClose } = props;
  const isChecked = selected.includes(item.id);
  const onSelectData = () => {
    onSelect?.(item.id);
    onClose?.();
  };
  return (
    <TouchableOpacity
      testID="renderItemDefault"
      onPress={onSelectData}
      disabled={isChecked}
      style={styles.itemContainer}
    >
      <Text numberOfLines={1} style={{ opacity: isChecked ? 0.5 : 1 }}>
        {item.title}
      </Text>
      <View style={styles.itemContent}>
        <FontAwesome5Icon
          name="check"
          size={18}
          color={isChecked ? colors.primary : colors.iconPopupColor}
          style={{ opacity: isChecked ? 0.5 : 1 }}
        />

        {haveUnSelect ? (
          <TouchableOpacity onPress={onUnSelect}>
            <FontAwesome6Icon
              name="xmark"
              size={22}
              color={isChecked ? colors.red : colors.iconPopupColor}
            />
          </TouchableOpacity>
        ) : null}
      </View>
    </TouchableOpacity>
  );
};

const createStyles = (colors: typeof defaultColors, size: typeof sizes) => {
  return StyleSheet.create({
    container: {
      flex: 1,
    },
    content: {
      flex: 1,
      paddingVertical: size.padding.sm,
    },
    itemContainer: {
      flex: 1,
      flexDirection: "row",
      justifyContent: "space-between",
      alignItems: "center",
      paddingVertical: size.padding.sm,
    },
    itemContent: {
      flexDirection: "row",
      alignItems: "center",
      gap: size.margin.lg,
      marginRight: size.margin.sm,
    },
    separator: {
      height: sizes.borderWidth.xs,
      width: "100%",
      backgroundColor: colors.separator,
      opacity: 0.5,
    },
    btnAddItem: {
      backgroundColor: colors.primary,
      padding: size.padding.sm,
      borderRadius: size.borderRadius.sm,
      justifyContent: "center",
      alignItems: "center",
      margin: size.margin.sm,
    },
    textAddItem: {
      color: colors.white,
      fontWeight: size.fontWeight.bold as "bold",
      fontSize: size.fontSize.lg,
    },
  });
};
