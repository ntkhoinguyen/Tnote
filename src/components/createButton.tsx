import React, { useMemo } from "react";
import { TouchableOpacity, StyleSheet } from "react-native";
import { Ionicons } from "@expo/vector-icons";

import { useAppContext } from "../useHook/useAppContext";
import { defaultColors } from "@/src/themes/colors";
import { sizes } from "@/src/themes/sizes";
type ButtonCreateType = {
  onPress: () => void;
  containerStyle?: object;
};

export const ButtonCreate: React.FC<ButtonCreateType> = (props) => {
  const { colors, sizes } = useAppContext();
  const styles = useMemo(() => createStyles(colors, sizes), [colors, sizes]);

  return (
    <TouchableOpacity
      testID="buttonCreate"
      onPress={props.onPress}
      style={[styles.container, props.containerStyle]}
    >
      <Ionicons name="add" size={32} color={colors.white} />
    </TouchableOpacity>
  );
};

const createStyles = (colors: typeof defaultColors, size: typeof sizes) => {
  return StyleSheet.create({
    container: {
      backgroundColor: colors.primary,
      padding: size.padding.sm,
      borderRadius: "50%",
      position: "absolute",
      right: 30,
      bottom: 30,
    },
  });
};

export default ButtonCreate;
