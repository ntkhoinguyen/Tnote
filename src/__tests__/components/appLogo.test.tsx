import { render } from "@testing-library/react-native";

import { AppLogo } from "@/src/components/appLogo";
import { sizes } from "@/src/themes/sizes";

describe("Test AppLogo Component", () => {
  it("renders with correct props", () => {
    const { getByTestId } = render(<AppLogo />);

    const logo = getByTestId("imgAppLogo");
    expect(logo.props.resizeMode).toBe("contain");
    expect(logo.props.style).toMatchObject([{
      width: sizes.appLogoWidth,
      height: sizes.appLogoHeight,
    }, { opacity: 1 }]);
  });

  test("snapshot testing for AppLogo", () => {
    const tree = render(<AppLogo />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
