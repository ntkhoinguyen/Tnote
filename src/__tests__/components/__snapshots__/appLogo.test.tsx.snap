// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test AppLogo Component snapshot testing for AppLogo 1`] = `
<Image
  resizeMode="contain"
  source={
    {
      "testUri": "../../../assets/images/logo.png",
    }
  }
  style={
    [
      {
        "height": 100,
        "width": 100,
      },
      {
        "opacity": 1,
      },
    ]
  }
  testID="imgAppLogo"
/>
`;
